File: src/mutual/match.py
"""Mutual — NSW（Nash Social Welfare）匹配求解 + envy 公平性检查。

对应 spec/02-stages.md §8、docs/engineering-plan.md §4.2（Phase 2）。

本模块是**无外部依赖**的独立求解器：只用 numpy（不依赖 FairRec/cvxpy/torch）。
输入是 :class:`~mutual.schemas.PrefMatrix`（双向偏好矩阵），输出为
``(list[Edge], match_prob, envy_report)``。

算法：按 NSW 分数（双向偏好几何平均 ``sqrt(pref_lr[i,j] * pref_rl[j,i])``）
降序的确定性贪心 b-matching（平局取 ``(i, j)`` 字典序）。

边界（spec/05-boundaries.md）：
- §7 度约束 ``b_max`` 绑定 member（左）侧；``pool_b_max`` 可选绑定 pool（右）侧。
  同集（left == right，如 cohort/full 模式）退化为无向图：``b_max`` 对称约束
  双方；``match_prob`` 对称存储。
- §3 未打分候选保留 embedding 权重（在 score 阶段已归一化，此处用偏好矩阵兜底）。

``b_min``（每人最少匹配数，qodo #9）：下界强制需要最小度 b-matching，
可行性依赖候选图密度（稀疏市场/novelty 排除后可能无解）。本求解器不静默
假装满足——贪心后做**显式可行性检查**，报告 ``b_min_violations``
（度数不足的 member id 列表）与 ``b_min_satisfied``，由调用方决定
（继续运行 / 报警 / 门禁阻断）。上界贪心保证不超 ``b_max``；下界
不可满足时**显式报告**，绝不静默吞掉。

说明：match 阶段只见 ``PrefMatrix``（无独立可分离的 embed/llm 分），因此
边的 ``final_weight`` 取 **NSW 分数**，``embed_score``/``llm_score`` 字段
以该分数重建（blending 公式在 embed==llm==nsw 时退化为 nsw）。
归一化由 :func:`mutual.score.prepare_normalized_scores` 在 score 阶段完成，
``reference_scores`` 在此仅透传。
"""

from __future__ import annotations

from math import sqrt
from typing import Any, List, Set, Tuple

import numpy as np

from .schemas import Edge, PrefMatrix, stable_pair_id


def solve_match(
    pref_matrix: PrefMatrix,
    matching_config: dict,
    blending_config: dict,
    reference_scores: np.ndarray | None = None,
) -> tuple[list[Edge], np.ndarray, dict]:
    """NSW 匹配求解 + envy 检查。

    Args:
        pref_matrix: 双向偏好矩阵（left→right 与 right→left）。
        matching_config: 度约束（``b_max``/``pool_b_max`` 上界贪心强制；
            ``b_min`` 下界显式可行性检查——不可满足时报告
            ``b_min_violations``，不静默，见模块 docstring）。
        blending_config: 混合权重（``embed_weight``/``llm_weight``）。
        reference_scores: 可选参考分数线；由 score 阶段消费，此处透传。

    Returns:
        ``(list[Edge], match_prob, envy_report)``：
        - 匹配边（按 ``(-final_weight, pair_id)`` 排序）；
        - 匹配概率矩阵（确定性匹配 → 0/1，shape ``[M, N]``；
          同集（无向）匹配对称存储 ``prob[i,j] == prob[j,i]``）；
        - envy 报告（``left_envy_count``/``right_envy_count``/``total_envy``/
          ``left``/``right``）+ ``b_min`` 可行性字段
          （``b_min``/``b_min_violations``/``b_min_satisfied``）。
    """
    M = len(pref_matrix.left_ids)
    N = len(pref_matrix.right_ids)
    b_min = max(0, _to_int(matching_config.get("b_min"), 0))
    if M == 0 or N == 0:
        report = _empty_envy()
        _attach_b_min_report(report, pref_matrix, np.zeros((M, N), dtype=int), b_min)
        return [], np.zeros((M, N), dtype=int), report

    b_max = _to_int(matching_config.get("b_max"), 0)
    pool_b_max = matching_config.get("pool_b_max")
    if pool_b_max is not None:
        pool_b_max = _to_int(pool_b_max, 0)

    w_embed = _to_float(blending_config.get("embed_weight"), 0.5)
    w_llm = _to_float(blending_config.get("llm_weight"), 0.5)

    same_set = list(pref_matrix.left_ids) == list(pref_matrix.right_ids)

    match_prob = np.zeros((M, N), dtype=int)
    matched_pairs: List[Tuple[int, int]] = []

    # 候选对按 NSW 分数降序（平局取 (i, j) 字典序）——全局互惠最优意图。
    candidates: List[Tuple[float, int, int]] = []
    if same_set:
        # 同一集合（如 cohort）：general matching（无向图，i < j 无序对）。
        for i in range(M):
            for j in range(i + 1, N):
                nsw = _nsw_score(pref_matrix, i, j)
                if nsw <= 0:
                    continue
                candidates.append((nsw, i, j))
        candidates.sort(key=lambda t: (-t[0], t[1], t[2]))
        deg = [0] * M
        for _nsw, i, j in candidates:
            if deg[i] < b_max and deg[j] < b_max:
                matched_pairs.append((i, j))
                match_prob[i, j] = 1
                match_prob[j, i] = 1  # 无向匹配：对称存储
                deg[i] += 1
                deg[j] += 1
    else:
        # 二部图（如 market/batch）：b_max 绑定 member（左）侧，
        # pool_b_max 可选绑定 pool（右）侧（spec/05-boundaries.md §7）。
        for i in range(M):
            for j in range(N):
                nsw = _nsw_score(pref_matrix, i, j)
                if nsw <= 0:
                    continue
                candidates.append((nsw, i, j))
        candidates.sort(key=lambda t: (-t[0], t[1], t[2]))
        left_deg = [0] * M
        right_deg = [0] * N
        for _nsw, i, j in candidates:
            left_ok = left_deg[i] < b_max
            right_ok = pool_b_max is None or right_deg[j] < pool_b_max
            if left_ok and right_ok:
                matched_pairs.append((i, j))
                match_prob[i, j] = 1
                left_deg[i] += 1
                right_deg[j] += 1

    edges = _build_edges(pref_matrix, matched_pairs, w_embed, w_llm)
    envy_report = check_envy(pref_matrix, match_prob)
    _attach_b_min_report(envy_report, pref_matrix, match_prob, b_min)
    return edges, match_prob, envy_report


def check_envy(pref_matrix: PrefMatrix, match_prob: np.ndarray) -> dict:
    """检查匹配结果中的 envy 公平性（own-best 语义，基于完整匹配集）。

    语义（与 :func:`mutual.evaluate.evaluate` 的 envy 计数一致）：
    左节点 ``i`` 嫉妒 ``i2`` ⟺ ``i2`` 的匹配集中存在 ``j2``，使
    ``pref_left_to_right[i, j2]`` 严格大于 ``i`` 自己**最优**匹配的偏好值。
    右侧同构（用 ``pref_right_to_left``）。

    Args:
        pref_matrix: 双向偏好矩阵。
        match_prob: 匹配概率矩阵（0/1，确定性匹配；同集匹配须对称存储）。

    Returns:
        ``{"left_envy_count", "right_envy_count", "total_envy", "left", "right"}``，
        其中 ``left``/``right`` 为 ``(envier, envied)`` 索引对的列表。
    """
    M = len(pref_matrix.left_ids)
    N = len(pref_matrix.right_ids)

    left_matches: List[Set[int]] = [
        {j for j in range(N) if match_prob[i, j] > 0.5} for i in range(M)
    ]
    right_matches: List[Set[int]] = [
        {i for i in range(M) if match_prob[i, j] > 0.5} for j in range(N)
    ]

    left_envy: List[Tuple[int, int]] = []
    for i, own in enumerate(left_matches):
        if not own:
            continue
        own_best = max(float(pref_matrix.pref_left_to_right[i, j]) for j in own)
        for i2, other in enumerate(left_matches):
            if i2 == i or not other:
                continue
            if any(float(pref_matrix.pref_left_to_right[i, j2]) > own_best for j2 in other):
                left_envy.append((i, i2))

    right_envy: List[Tuple[int, int]] = []
    for j, own in enumerate(right_matches):
        if not own:
            continue
        own_best = max(float(pref_matrix.pref_right_to_left[j, i]) for i in own)
        for j2, other in enumerate(right_matches):
            if j2 == j or not other:
                continue
            if any(float(pref_matrix.pref_right_to_left[j, i2]) > own_best for i2 in other):
                right_envy.append((j, j2))

    return {
        "left_envy_count": len(left_envy),
        "right_envy_count": len(right_envy),
        "total_envy": len(left_envy) + len(right_envy),
        "left": left_envy,
        "right": right_envy,
    }


# ---------------------------------------------------------------------------
# 内部 helper（纯函数）
# ---------------------------------------------------------------------------


def _nsw_score(pref_matrix: PrefMatrix, i: int, j: int) -> float:
    """NSW 分数：双向偏好的几何平均（``sqrt(pref_lr * pref_rl)``）。"""
    a = float(pref_matrix.pref_left_to_right[i, j])
    b = float(pref_matrix.pref_right_to_left[j, i])
    return sqrt(a * b)


def _build_edges(
    pref_matrix: PrefMatrix,
    matched_pairs: List[Tuple[int, int]],
    w_embed: float,
    w_llm: float,
) -> List[Edge]:
    """把匹配对构造成 :class:`~mutual.schemas.Edge`。

    边方向：左节点 ``left_ids[i]`` × 右节点 ``right_ids[j]``（batch 模式
    即 member × pool）。
    ``final_weight``：blending 权重混合；因 match 阶段仅见偏好矩阵，
    embed/llm 均以 NSW 分数重建（混合退化为 ``(w_embed + w_llm) * nsw``）。
    """
    edges: List[Edge] = []
    for i, j in matched_pairs:
        user1 = pref_matrix.left_ids[i]
        user2 = pref_matrix.right_ids[j]
        a_to_b = float(pref_matrix.pref_left_to_right[i, j])
        b_to_a = float(pref_matrix.pref_right_to_left[j, i])
        nsw = sqrt(a_to_b * b_to_a)
        final_weight = (w_embed + w_llm) * nsw
        edges.append(
            Edge(
                user1=user1,
                user2=user2,
                pair_id=stable_pair_id(user1, user2),
                final_weight=final_weight,
                embed_score=nsw,
                llm_score=nsw,
                embed_score_normalized=None,
                llm_score_normalized=None,
                llm_score_a_to_b=a_to_b,
                llm_score_b_to_a=b_to_a,
                intro="",
                starter_topics="",
            )
        )
    edges.sort(key=lambda e: (-e.final_weight, e.pair_id))
    return edges


def _empty_envy() -> dict:
    return {
        "left_envy_count": 0,
        "right_envy_count": 0,
        "total_envy": 0,
        "left": [],
        "right": [],
    }


def _attach_b_min_report(
    report: dict,
    pref_matrix: PrefMatrix,
    match_prob: np.ndarray,
    b_min: int,
) -> None:
    """把 ``b_min`` 可行性字段附加进报告（qodo #9，原地修改）。

    度数口径：member 侧 = 左节点（二部图）或全部节点（同集无向图，
    此时所有节点都是 member，度数按 ``match_prob`` 行和计——对称存储下
    行和即总度数）。

    ``b_min <= 0`` 时不启用（报告仍写入 ``b_min`` 值与空 violations，
    保持报告 shape 稳定，便于下游消费）。
    """
    M = len(pref_matrix.left_ids)
    N = len(pref_matrix.right_ids)
    same_set = list(pref_matrix.left_ids) == list(pref_matrix.right_ids)

    if same_set:
        degrees = {pref_matrix.left_ids[i]: int(match_prob[i].sum()) for i in range(M)}
    else:
        degrees = {
            pref_matrix.left_ids[i]: int(match_prob[i, :].sum()) for i in range(M)
        }
    # N == 0 时 match_prob 形状 [M, 0]，行和为 0——全部 member 违约，正确。
    violations = [uid for uid, deg in degrees.items() if deg < b_min]
    report["b_min"] = b_min
    report["b_min_violations"] = violations
    report["b_min_satisfied"] = not violations


def _to_int(value: Any, default: int) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return default


def _to_float(value: Any, default: float) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return default

File: internal/engine/match.go
package engine

import (
	"math"
	"sort"

	"github.com/Cloudbird-Software/mutual/internal/domain"
)

// MatchingConfig 是度约束（config["matching"]）。
type MatchingConfig struct {
	// BMin: 每人最少匹配数（下界显式可行性检查，qodo #9）。
	BMin int
	// BMax: 每人最多匹配数（上界贪心强制；绑定 member 左侧，
	// 同集模式对称约束双方）。
	BMax int
	// PoolBMax: pool（右）侧度数上界；nil = 不限。
	PoolBMax *int
}

// BlendingConfig 是 embed/llm 分数混合权重（config["blending"]）。
type BlendingConfig struct {
	EmbedWeight float64
	LLMWeight   float64
}

// MatchOutcome 是 SolveMatch 的输出三件套。
type MatchOutcome struct {
	// Edges: 匹配边（按 (-final_weight, pair_id) 排序）。
	Edges []domain.Edge
	// MatchProb: 匹配概率矩阵（确定性匹配 → 0/1，[M][N]；
	// 同集（无向）匹配对称存储 prob[i][j] == prob[j][i]）。
	MatchProb domain.Matrix
	// EnvyReport: envy 公平性报告 + b_min 可行性字段。
	EnvyReport map[string]any
}

// SolveMatch NSW 匹配求解 + envy 公平性检查（match 阶段，纯函数）。
//
// 算法：按 NSW 分数（双向偏好几何平均 sqrt(pref_lr·pref_rl)）降序的
// 确定性贪心 b-matching（平局取 (i,j) 字典序）——全局互惠最优意图。
//
// 边界（spec/05-boundaries.md）：
//   - §7 度约束 BMax 绑定 member（左）侧；PoolBMax 可选绑定 pool 侧；
//     同集（left == right）退化为无向图：BMax 对称约束双方。
//   - §3 未打分候选已在 pre_matrix 用 embed 权重兜底。
//   - BMin 不可满足时显式报告（b_min_violations），绝不静默吞掉；
//     由调用方决定继续运行 / 报警 / 门禁阻断。
//
// 说明：match 阶段只见 PrefMatrix（无独立可分离的 embed/llm 分），
// 因此边的 final_weight 取 NSW 分数（blending 在 embed==llm==nsw 时
// 退化为 (w_embed + w_llm)·nsw）；归一化已在 score 阶段完成。
func SolveMatch(prefMatrix *domain.PrefMatrix, matching MatchingConfig, blending BlendingConfig) MatchOutcome {
	m := prefMatrix.M()
	n := prefMatrix.N()
	bMin := matching.BMin
	if bMin < 0 {
		bMin = 0
	}
	matchProb := domain.NewMatrixZeros(m, n)

	if m == 0 || n == 0 {
		report := emptyEnvy()
		attachBMinReport(report, prefMatrix, matchProb, bMin)
		return MatchOutcome{Edges: nil, MatchProb: matchProb, EnvyReport: report}
	}

	poolBMax := matching.PoolBMax
	sameSet := sameIDList(prefMatrix.LeftIDs, prefMatrix.RightIDs)
	// BMax <= 0 显式定义为"不限度数"（上界 m+n）——否则配置漏写/拼错
	// b_max 会静默产出零匹配，且 envy=0 看起来像"完美公平"（CodeRabbit）。
	// Python 基线同样默认 0 但无此防护；golden 配置显式 b_max=4，不受影响。
	bMax := matching.BMax
	if bMax <= 0 {
		bMax = m + n
	}

	var matchedPairs []pairIJ

	// 候选对按 NSW 分数降序（平局取 (i, j) 字典序）。
	type cand struct {
		nsw  float64
		i, j int
	}
	var candidates []cand
	if sameSet {
		// 同集（cohort/full）：general matching，i < j 无序对。
		for i := 0; i < m; i++ {
			for j := i + 1; j < n; j++ {
				nsw := nswScore(prefMatrix, i, j)
				if nsw <= 0 {
					continue
				}
				candidates = append(candidates, cand{nsw: nsw, i: i, j: j})
			}
		}
	} else {
		// 二部图（market/batch）：全有序对。
		for i := 0; i < m; i++ {
			for j := 0; j < n; j++ {
				nsw := nswScore(prefMatrix, i, j)
				if nsw <= 0 {
					continue
				}
				candidates = append(candidates, cand{nsw: nsw, i: i, j: j})
			}
		}
	}
	sort.SliceStable(candidates, func(a, b int) bool {
		x, y := candidates[a], candidates[b]
		if x.nsw != y.nsw {
			return x.nsw > y.nsw
		}
		if x.i != y.i {
			return x.i < y.i
		}
		return x.j < y.j
	})

	if sameSet {
		deg := make([]int, m)
		for _, c := range candidates {
			if deg[c.i] < bMax && deg[c.j] < bMax {
				matchedPairs = append(matchedPairs, pairIJ{c.i, c.j})
				matchProb[c.i][c.j] = 1
				matchProb[c.j][c.i] = 1 // 无向匹配：对称存储
				deg[c.i]++
				deg[c.j]++
			}
		}
	} else {
		leftDeg := make([]int, m)
		rightDeg := make([]int, n)
		for _, c := range candidates {
			leftOK := leftDeg[c.i] < bMax
			rightOK := poolBMax == nil || rightDeg[c.j] < *poolBMax
			if leftOK && rightOK {
				matchedPairs = append(matchedPairs, pairIJ{c.i, c.j})
				matchProb[c.i][c.j] = 1
				leftDeg[c.i]++
				rightDeg[c.j]++
			}
		}
	}

	edges := buildEdges(prefMatrix, matchedPairs, blending.EmbedWeight, blending.LLMWeight)
	envyReport := CheckEnvy(prefMatrix, matchProb)
	attachBMinReport(envyReport, prefMatrix, matchProb, bMin)
	return MatchOutcome{Edges: edges, MatchProb: matchProb, EnvyReport: envyReport}
}

// CheckEnvy 检查匹配结果中的 envy 公平性（own-best 语义）。
//
// 语义（与 Evaluate 的 envy 计数逐位一致，改语义 = 改 oracle）：
// 左节点 i 嫉妒 i2 ⟺ i2 的匹配集中存在 j2，使 pref_lr[i][j2]
// 严格大于 i 自己最优匹配的偏好值。右侧同构。配对计算复用
// evaluate.go 的 collectRowMatches / envyPairs，两侧口径不分叉。
func CheckEnvy(prefMatrix *domain.PrefMatrix, matchProb domain.Matrix) map[string]any {
	// 左侧：envier = 行；右侧行视角取转置（行 = right 侧）。
	leftMatches := collectRowMatches(matchProb)
	rightMatches := collectRowMatches(transpose(matchProb))

	leftEnvy := envyPairs(prefMatrix.PrefLeftToRight, leftMatches)
	rightEnvy := envyPairs(prefMatrix.PrefRightToLeft, rightMatches)

	leftList := make([]any, len(leftEnvy))
	for k, p := range leftEnvy {
		leftList[k] = []int{p[0], p[1]}
	}
	rightList := make([]any, len(rightEnvy))
	for k, p := range rightEnvy {
		rightList[k] = []int{p[0], p[1]}
	}
	return map[string]any{
		"left_envy_count":  len(leftEnvy),
		"right_envy_count": len(rightEnvy),
		"total_envy":       len(leftEnvy) + len(rightEnvy),
		"left":             leftList,
		"right":            rightList,
	}
}

// nswScore NSW 分数：双向偏好的几何平均 sqrt(pref_lr·pref_rl)。
//
// 非正乘积（含异号乘积的 NaN 风险）返回 0：Python 基线此处是
// math.sqrt 对负数直接抛 ValueError，Go 侧以 0 优雅降级（该候选
// 不参与匹配），golden 路径乘积恒正、逐位一致不受影响。NaN 若
// 放行会绕过调用侧 nsw <= 0 过滤并破坏排序的严格弱序（CodeRabbit）。
func nswScore(pm *domain.PrefMatrix, i, j int) float64 {
	a := pm.PrefLeftToRight[i][j]
	b := pm.PrefRightToLeft[j][i]
	p := a * b
	if p <= 0 || math.IsNaN(p) || math.IsInf(p, 0) {
		return 0
	}
	return math.Sqrt(p)
}

// pairIJ 是偏好矩阵中的有序位置对（左行 i，右列 j）。
type pairIJ struct{ i, j int }

// buildEdges 把匹配对构造成 Edge（final_weight = blending 混合；
// 因 match 只见偏好矩阵，embed/llm 均以 NSW 重建）。
func buildEdges(pm *domain.PrefMatrix, matchedPairs []pairIJ, wEmbed, wLLM float64) []domain.Edge {
	edges := make([]domain.Edge, 0, len(matchedPairs))
	for _, p := range matchedPairs {
		user1 := pm.LeftIDs[p.i]
		user2 := pm.RightIDs[p.j]
		aToB := pm.PrefLeftToRight[p.i][p.j]
		bToA := pm.PrefRightToLeft[p.j][p.i]
		nsw := nswScore(pm, p.i, p.j) // 与候选过滤同一实现，避免两处逻辑分叉
		finalWeight := (wEmbed + wLLM) * nsw
		edges = append(edges, domain.Edge{
			User1:        user1,
			User2:        user2,
			PairID:       domain.StablePairID(user1, user2),
			FinalWeight:  finalWeight,
			EmbedScore:   nsw,
			LLMScore:     nsw,
			LLMScoreAToB: &aToB,
			LLMScoreBToA: &bToA,
		})
	}
	sort.SliceStable(edges, func(a, b int) bool {
		if edges[a].FinalWeight != edges[b].FinalWeight {
			return edges[a].FinalWeight > edges[b].FinalWeight
		}
		return edges[a].PairID < edges[b].PairID
	})
	return edges
}

func emptyEnvy() map[string]any {
	return map[string]any{
		"left_envy_count":  0,
		"right_envy_count": 0,
		"total_envy":       0,
		"left":             []any{},
		"right":            []any{},
	}
}

// attachBMinReport 把 b_min 可行性字段附加进报告（原地修改）。
//
// 度数口径：member 侧 = 左节点（二部图）或全部节点（同集无向图，
// 对称存储下行和即总度数）。b_min <= 0 时不启用，但字段仍写入
// （报告 shape 稳定，便于下游消费）。
func attachBMinReport(report map[string]any, pm *domain.PrefMatrix, matchProb domain.Matrix, bMin int) {
	m := pm.M()

	var violations []string
	for i := 0; i < m; i++ {
		// 度数 = 行和（同集模式对称存储下行和即总度数；二部图
		// 模式即 member 侧度数——两种口径在此统一）。
		degree := 0
		for _, v := range matchProb[i] {
			if v > matchThreshold {
				degree++
			}
		}
		if degree < bMin {
			violations = append(violations, string(pm.LeftIDs[i]))
		}
	}
	if violations == nil {
		violations = []string{}
	}
	report["b_min"] = bMin
	report["b_min_violations"] = violations
	report["b_min_satisfied"] = len(violations) == 0
}

func sameIDList(a, b []domain.UserID) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
